contracts/non-fungible-tokens/andromeda-marketplace/src/contract.rs

```diff
@@ -0,0 +1,946 @@
+        deps.api,
+        info,
+        BaseInstantiateMsg {
+            ado_type: "marketplace".to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            operators: None,
+            modules: msg.modules,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+
+    contract.module_hook::<Response>(
+        deps.storage,
+        deps.api,
+        deps.querier,
+        AndromedaHook::OnExecute {
+            sender: info.sender.to_string(),
+            payload: encode_binary(&msg)?,
+        },
+    )?;
+    match msg {
+        ExecuteMsg::AndrReceive(msg) => {
+            ADOContract::default().execute(deps, env, info, msg, execute)
+        }
+        ExecuteMsg::ReceiveNft(msg) => handle_receive_cw721(deps, env, info, msg),
+        ExecuteMsg::UpdateSale {
+            token_id,
+            token_address,
+            coin_denom,
+            whitelist,
+            price,
+        } => execute_update_sale(
+            deps,
+            env,
+            info,
+            token_id,
+            token_address,
+            price,
+            coin_denom,
+            whitelist,
+        ),
+        ExecuteMsg::Buy {
+            token_id,
+            token_address,
+        } => execute_buy(deps, env, info, token_id, token_address),
+        ExecuteMsg::CancelSale {
+            token_id,
+            token_address,
+        } => execute_cancel(deps, env, info, token_id, token_address),
+    }
+}
+
+fn handle_receive_cw721(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    match from_binary(&msg.msg)? {
+        Cw721HookMsg::StartSale {
+            price,
+            coin_denom,
+            whitelist,
+        } => execute_start_sale(
+            deps,
+            env,
+            msg.sender,
+            msg.token_id,
+            info.sender.to_string(),
+            price,
+            coin_denom,
+            whitelist,
+        ),
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_start_sale(
+    deps: DepsMut,
+    _env: Env,
+    sender: String,
+    token_id: String,
+    token_address: String,
+    price: Uint128,
+    coin_denom: String,
+    whitelist: Option<Vec<Addr>>,
+) -> Result<Response, ContractError> {
+    // Price can't be zero
+    require(price > Uint128::zero(), ContractError::InvalidZeroAmount {})?;
+
+    let sale_id = get_and_increment_next_sale_id(deps.storage, &token_id, &token_address)?;
+
+    let whitelist_str = format!("{:?}", &whitelist);
+
+    TOKEN_SALE_STATE.save(
+        deps.storage,
+        sale_id.u128(),
+        &TokenSaleState {
+            coin_denom: coin_denom.clone(),
+            sale_id,
+            whitelist,
+            owner: sender,
```

Comment:
Is the sender the contract or the token owner?

Reply:
The token owner. That's what `fn assert_sale_created` shows